rs-backend/src/api.rs

```diff
@@ -0,0 +1,167 @@
+        /// Bootstrap counts for twice as much that all other benchmarks do.
+        /// First week in vector is the current week.
+        pub summaries: Vec<Percent>,
+
+        /// 12 week long list of dates from which data was used for that week.
+        /// First week in vector is the current week.
+        pub dates: Vec<String>,
+    }
+}
+
+pub mod info {
+    #[derive(Serialize)]
+    pub struct Response {
+        /// Sorted vector of crate names (benchmarks and all crates in the rust distribution)
+        pub crates: Vec<String>,
+
+        /// Sorted vector of phase names.
+        pub phases: Vec<String>,
+
+        /// Sorted vector of benchmark names.
+        pub benchmarks: Vec<String>,
+    }
+}
+
+pub mod data {
+    use serde_json::Value;
+
+    use load::Kind;
+    use server::{GroupBy, OptionalDate};
+
+    #[derive(Serialize, Deserialize)]
+    pub struct Request {
+        #[serde(rename="start")]
+        pub start_date: OptionalDate,
+
+        #[serde(rename="end")]
+        pub end_date: OptionalDate,
+        pub kind: Kind,
+        pub group_by: GroupBy,
+
+        /// Which crates to return data for
+        pub crates: Vec<String>,
+
+        /// Which phases to return data for
+        pub phases: Vec<String>,
+    }
+
+    /// The response consists of a list of objects, from oldest to newest,
+    /// with these fields:
+    ///   - commit: Git commit hash of the compiler these results were obtained with.
+    ///   - date: The date of this run; in the JS_DATE_FORMAT (TODO: Link/reference)
+    ///   - data: An object with keys being the requested crates/phases (depending on group_by);
+    ///       - rss: u64 of memory usage (TODO: What unit? MB?)
+    ///       - time: f64 of duration for compiling (TODO: Is this in seconds?)
+    // TODO: Use a struct instead of Value.
+    #[derive(Serialize, Deserialize)]
+    pub struct Response(pub Vec<Value>);
+}
+
+pub mod tabular {
+    use std::collections::HashMap;
+
+    use load::{Kind, Timing};
+    use server::OptionalDate;
+
+    #[derive(Serialize, Deserialize)]
+    pub struct Request {
+        pub kind: Kind,
+        pub date: OptionalDate,
+    }
+
+    #[derive(Serialize, Deserialize)]
+    pub struct Response {
+        pub commit: String,
+
+        /// Mapping of all crates timed on this date to their timings by phase
+        pub data: HashMap<String, HashMap<String, Timing>>,
+        pub date: String,
+    }
+}
+
+pub mod days {
+    use serde_json::Value;
+
+    use load::Kind;
+    use server::{OptionalDate, GroupBy};
+
+    #[derive(Serialize, Deserialize)]
+    pub struct Request {
+        pub kind: Kind,
+        pub group_by: GroupBy,
+        pub dates: Vec<OptionalDate>,
+
+        /// Which crates to return data for
+        pub crates: Vec<String>,
+
+        /// Which phases to return data for
+        pub phases: Vec<String>,
+    }
+
+    #[derive(Serialize, Deserialize)]
+    pub struct Response(pub Vec<Value>);
```

Comment:
The main reason is that the helper function `get_data_for_date` returns a value currently. I plan to rewrite it at some point in the future (read: soon) to not do that, and then both of the value instances here will become nicer.